File: exercises/C/DerivedDataTypes/pingpong-derdaty.c
//
//  Program in which 2 processes repeatedly pass M columns of a matrix back and forth
//
// The number/array swapped changes from swap to swap; from back and forth.
//

#include <stdio.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <mpi.h>

#define N  4    // For an NxN matrix
#define M  1    // Number of columns to be sent together
#define ColStart  2    // where ColStart is the start column
#define itmax 4 // Number of iterations = itmax - 1

int main(int argc, char* argv[]){
  int rank, size, ierr,length;
  int start,stop,i,j,iter;    
  double rbuffer,sbuffer;
  MPI_Comm comm;
  MPI_Status status;

  double sendmatrix[N][N], recvmatrix[N][N];
  MPI_Datatype mcoltype;


  comm  = MPI_COMM_WORLD;

  ierr = MPI_Init(&argc, &argv);
  MPI_Comm_rank(comm, &rank);            
  MPI_Comm_size(comm, &size); 

  // Abort if not run on 2 processors.
  if(size != 2){
    if(rank == 0){
      printf("The code can only be run on 2 processors.\n");
    }
    MPI_Finalize();
    exit(1);
  }

  // Number of rows must not exceed number of coulumns, and start row to be sent must not exceed N.
  if(M > N || ColStart > N ||(ColStart+M) > (N+1)){
    printf("The number of rows MUST NOT be bigger than the number of coulumns: M=%d, N=%d, ColStart=%d \n",M,N,ColStart);
    MPI_Finalize();
    exit(1);
  }

  // Create new datatype mrowtype to send M columns and commit it.
  MPI_Type_vector(N,M,N,MPI_DOUBLE,&mcoltype);
  MPI_Type_commit(&mcoltype);

  // Initialize matrix
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      sendmatrix[i][j] = (i+1)*pow(10,j)+rank;
      printf("Matrix intialized on rank %d: i=%d, j=%d, element=%f \n",rank,i,j,sendmatrix[i][j]);
    }
  }

  //E.g. sending columns 4, 5 and 6 of a matrix 10x10 back and forth between rank 0 and rank 1
  // corresponds to ColStart = 4, M = 3, N = 10

  for(iter=1;iter<itmax;iter++){
   if(rank == 0){

    for(i=0;i<N;i++){
     for(j=ColStart;j<(ColStart+M);j++){
     }
    }

    MPI_Ssend(&(sendmatrix[0][ColStart-1]),1,mcoltype,1,50,comm);
    MPI_Recv(&(recvmatrix[0][ColStart-1]),1,mcoltype,1,60,comm,&status);
   }
   else{

    for(i=0;i<N;i++){
     for(j=ColStart;j<(ColStart+M);j++){
     }
    }

    MPI_Recv(&(recvmatrix[0][ColStart-1]),1,mcoltype,0,50,comm,&status);
    MPI_Ssend(&(sendmatrix[0][ColStart-1]),1,mcoltype,0,60,comm);

   }

   for(i=0;i<N;i++){
    for(j=ColStart;j<(ColStart+M);j++){
     sendmatrix[i][j-1] = recvmatrix[i][j-1];
    }
   }
  }


  // Matrix after itmax ping-pongs                                                                                   
  for(i=0;i<N;i++){
   for(j=0;j<N;j++){
    printf("Matrix in the end on rank %d: i=%d, j=%d, element=%f \n",rank,i,j,sendmatrix[i][j]);
   }
  }

  if(rank == 0){
    printf("Sending %d column(s) of an %d x %d matrix, starting with column %d and sending it %d times.\n",M,N,N,ColStart,itmax-1);
  }

  MPI_Finalize();

}

File: exercises/Fortran90/DerivedDataTypes/pingpong-derdaty.f90
!
! Program in which 2 processes repeatedly pass a M rows of a NxN matrix back and forth
!
!

program pingpong
implicit none
include 'mpif.h'

integer :: ierr, size, rank, comm, i, length, j
integer :: status(MPI_STATUS_SIZE)
real :: buffer(100000)
integer, parameter :: iter =3      ! Number of ping pongs to be executed
integer, parameter :: M=2,N=4,ColStart=2  ! M number of rows to be sent together, N number of coulumns, ColStart start row
integer :: mrowtype                ! new type to be created
double precision, dimension(N,N) :: sendmatrix, recvmatrix

comm = MPI_COMM_WORLD

call MPI_INIT(ierr)
call MPI_COMM_RANK(comm,rank,ierr)
call MPI_COMM_SIZE(comm,size,ierr)

! Code  must only run on 2 processors.
if(size.ne.2)then
 if(rank.eq.0) write(*,*) ' The code can only be run on 2 processors.'
 call MPI_FINALIZE(ierr)
 stop
endif

! Number of rows must not exceed number of coulumns, and start row to be sent must not exceed N.
if(M.gt.N.or.ColStart.gt.N.or.(ColStart+M).gt.(N+1))then
 write(*,*) 'The number of rows MUST NOT be bigger than the number of coulumns: ',M, N, ColStart
 call MPI_FINALIZE(ierr)
 stop
endif

! Create new datatype mrowtype to send M rows and commit it.
call MPI_TYPE_VECTOR(N,M,N,MPI_DOUBLE_PRECISION,mrowtype,ierr)
call MPI_TYPE_COMMIT(mrowtype,ierr)

! Initialize matrix.
do j = 1, N
 do i = 1, N
  sendmatrix(i,j) = i*10**(j-1)+rank
  write(*,*)'Intitialized matrix on rank ',rank,':',i,j, sendmatrix(i,j)
 enddo
enddo

! E.g. sending rows 4, 5 and 6 of a matrix back and forth between rank 0 and rank 1
! corresponds to ColStart = 4, M = 3

do i = 1, iter
 if (rank.eq.0)then

  call MPI_SSEND(sendmatrix(ColStart,1),1,mrowtype,1,50,comm,ierr)
  call MPI_RECV(recvmatrix(ColStart,1),1,mrowtype,1,60,comm,status,ierr)

 else

  call MPI_RECV(recvmatrix(ColStart,1),1,mrowtype,0,50,comm,status,ierr)
  call MPI_SSEND(sendmatrix(ColStart,1),1,mrowtype,0,60,comm,ierr)

 endif

 sendmatrix(ColStart:ColStart+M,1:N) = recvmatrix(ColStart:ColStart+M,1:N)

enddo

! Matrix after iter ping-pongs
do j=1,N
 do i=1,N
  write(*,*) 'Matrix in the end on rank ',rank,':', i,j,sendmatrix(i,j)
 enddo
enddo

if(rank.eq.0)then
 write(*,30) M,N,N,ColStart,iter
endif

30 format (1x, 'Sending ', i3 ,'  column(s) of an ', i3, ' x ', i3, ' matrix, starting with column ', i3 ,' and sending it ',i6,' times.')

call MPI_FINALIZE(ierr)

end program pingpong
